### PR TITLE
Merge conflicting `wp.editor` objects into a single, non-conflicting module

### PR DESCRIPTION
**Pull request by adamziel**
## Description
Solves https://github.com/WordPress/gutenberg/issues/33203

Situation: `packages/editor` is exposed as `window.wp.editor` in wp-admin

Problem: there is quite some code expecting to see a different (legacy) object under `window.wp.editor`

Proposed solution: assign all the members of legacy window.wp.editor to the new window.wp.editor like this: 
```
'Object.assign(window.wp.editor, window.wp.oldEditor);',
```

This PR is a Gutenberg counterpart of https://github.com/WordPress/wordpress-develop/pull/1481 in core

## How has this been tested?
1. Activate classic widgets plugins.
1. Insert text widget
1. Add some text to widget.
1. Deactivate classic widgets plugins.
1. Go to widget screen.
1. There is an error message if this PR is not applied.
1. There are no error messages if this PR is applied.

## Screenshots

124498441-3f40df80-ddb4-11eb-99e1-581b287f96de

(Credits for the screenshot and the test plan to @spacedmonkey)

## Types of changes
Bug fix (non-breaking change which fixes an issue)

### CODE REVIEW

**Comment by spacedmonkey** [comment]
This will like fix other issues, where other plugins have called `wp.editor.initialize`.
 See https://wpdirectory.net/search/01F9YAKV5H5WWHMADCK7VDDA1T

**Comment by kevin940726** [comment]
I fear that this approach is not ideal. If I'm understanding this correctly, it basically means that the gutenberg `editor` package will depend on `oldEditor` in core. We can't have `initialize` as a export member in the `editor` package anymore for instance.

I wonder, instead, if we could monkey-patch/proxy `wp.editor` somehow when it's running inside a legacy widget? Only then we'll return `wp.oldEditor`. Note that it should also take asynchronous code into account. Not sure if it'll work or if it can be done though.

File: packages/editor/src/index.js [hunk 1]
```diff
@@ -7,6 +7,13 @@ export { storeConfig, store } from './store';
 export * from './components';
 export * from './utils';
 
+// This module is exposed as window.wp.editor
+// Problem: there is quite some code expecting another, legacy object available under window.wp.editor
+// Solution: export all the members of legacy window.wp.editor from this new module to maintain backward compatibility
+// For more context, see https://github.com/WordPress/gutenberg/issues/33203
+export const { autop, removep, initialize, remove, getContent } =
```
Comment (kevin940726) [inline packages/editor/src/index.js]: I didn't even know this is valid JS 🤯

**Comment by noisysocks** [comment]
It's regrettable that we ever added `@wordpress/editor` since that name was already "taken" 😀

My only issue with this approach is that it means we're adding five functions to `@wordpress/editor` which we'll need to support forever. Maybe that's OK, I'm not sure. Perhaps we should fire `deprecated()` and encourage folks to use `wp.classicEditor` instead? I'm not sure. What do you think @youknowriad?

**Comment by noisysocks** [comment]
There’s also this similar issue with the `wp-edit-post` style dep: https://core.trac.wordpress.org/ticket/53569

Makes me wonder if a “fix” could be to add code to Core which fires off `doing_it_wrong()` if you enqueue `wp-editor` or `wp-edit-post` in the widgets screen.

**Comment by adamziel** [comment]
> Perhaps we should fire deprecated() and encourage folks to use wp.classicEditor instead?
> Makes me wonder if a “fix” could be to add code to Core which fires off doing_it_wrong() if you enqueue wp-editor or wp-edit-post in the widgets screen.

Both deprecated and doing_it_wrong sounds great to me, let's absolutely add that. Still, we need to make sure the existing code keeps working for another version or two as we start deprecating things.

A slightly different approach here would be to add another inline script like the one that creates `wp.oldEditor`:
```
    // Loading the old editor and its config to ensure the classic block works as expected.
    $scripts->add_inline_script(
        'editor',
        'window.wp.oldEditor = window.wp.editor;',
        'after'
```
And use it to bolt these five functions on top of the “new” wp.editor immediately after it’s loaded.

**Review by youknowriad — COMMENTED** [review]
Feels like this should be an inline script to `wp-editor` instead of something in the package directly because it's very WP specific, wdyt?

**Comment by adamziel** [comment]
Yup, let's try that – I'll publish a new PR soon

**Comment by adamziel** [comment]
Here's the inline-script approach: https://github.com/WordPress/wordpress-develop/pull/1481

**Comment by youknowriad** [comment]
@adamziel that looks good to me, should we do the same in Gutenberg.

**Comment by adamziel** [comment]
@youknowriad my intuition says no because maybe then we'd run both code paths and end up having two inline scripts enqueued; also in Gutenberg there's no `window.wp.oldEditor = window.wp.editor;` assignment, that part (and the original wp.editor) seems to be specific to core. What do you think?

**Comment by youknowriad** [comment]
@adamziel we do support WP 5.7 and WP 5.6 for now in Gutenberg though and when 5.8 get released, we'd shift to support 5.7 and 5.8.

**Comment by adamziel** [comment]
@youknowriad  Good point! I'll start a new PR for Gutenberg then

**Comment by adamziel** [comment]
@youknowriad I just updated this PR to use the inline script approach instead. In my testing this snippet is not being executed together with the one supplied by core because in 5.7 the widgets editor is provided via the plugin, and in 5.8 the plugin code paths don't seem to be executed. But even if there's a case where we'd run both, it doesn't really matter.

File: lib/widgets-page.php [hunk 2]
```diff
@@ -76,6 +76,7 @@ function gutenberg_widgets_init( $hook ) {
 		return;
 	}
 
+	gutenberg_enqueue_wp_old_editor_compatibility_snippet();
```
Comment (youknowriad) [inline lib/widgets-page.php]: Do you think this should be specific to these screens or always applied (for safety). If the latter, we have a dedicate place for this kind of inline scripts in Gutenberg, the `gutenberg_override_script` function

Comment (noisysocks) [reply lib/widgets-page.php]: Putting it in `gutenberg_override_script` makes sense to me. The post block editor might benefit from this compatibility shim too. At the very least it means there's more consistency.

**Review by draganescu — APPROVED** [review]
LGTM! Thank you folks for the great collab, this has been hunting us for a while now 👏🏻

**Review by youknowriad — APPROVED** [review]

File: packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js [hunk 3]
```diff
@@ -39,8 +39,6 @@ describe( 'WP Editor Meta Boxes', () => {
 			( textarea ) => textarea.value
 		);
 
-		expect( content ).toMatchInlineSnapshot(
-			`"<p>Typing in a metabox</p>"`
-		);
+		expect( content ).toBe( 'Typing in a metabox' );
```
Comment (youknowriad) [inline packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js:56]: Is this change expected, I feel this test has been updated back and forth several times already in the last couple releases, do we understand why this change is needed?

**Comment by noisysocks** [comment]
We don't need to backport this PR as there's already a commit awaiting backport here: https://core.trac.wordpress.org/ticket/53437

File: packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js (hunk 3 above)
Comment (adamziel) [reply packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js:56]: It turns out that TinyMCE's WordPress plugin does the following:
https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/_enqueues/vendor/tinymce/plugins/wordpress/plugin.js#L15
https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/_enqueues/vendor/tinymce/plugins/wordpress/plugin.js#L132-L133
https://github.com/WordPress/wordpress-develop/blob/2382765afa36e10bf3c74420024ad4e85763a47c/src/js/_enqueues/vendor/tinymce/plugins/wordpress/plugin.js#L593-L609

```js
hasWpautop = ( wp && wp.editor && wp.editor.autop && editor.getParam( 'wpautop', true ) ),
editor.on( 'BeforeSetContent', function( event ) {
	// ...
	if ( hasWpautop ) {
	    event.content = wp.editor.autop( event.content );
	} else {
	    // Prevent creation of paragraphs out of multiple HTML comments
	    event.content = event.content.replace( /-->\s+<!--/g, '--><!--' );
	}
});

// ...

editor.on( 'SaveContent', function( event ) {
	// ...
	// Keep empty paragraphs :(
	event.content = event.content.replace( /<p>(?:<br ?\/?>|\u00a0|\uFEFF| )*<\/p>/g, '<p>&nbsp;</p>' );

	if ( hasWpautop ) {
		event.content = wp.editor.removep( event.content );
	} else {
		// Restore formatting of block boundaries.
		event.content = event.content.replace( /-->\s*<!-- wp:/g, '-->\n\n<!-- wp:' );
	}
});
```



So:
- if `Object.assign( window.wp.editor, window.wp.oldEditor );` is in place, then hasWpauto is set to true by editor.getParam( 'wpautop', true ); .
	- When I switch from HTML to visual mode of TinyMCE, event.content is wrapped in a `<p> `
	- When I switch back from the visual mode to the HTML mode, SaveContent is dispatched and it calls wp.editor.removep which removes the `<p>` tags.
- If Object.assign thing is not in place, then hasWpauto is set to false. 
	- When I switch from HTML to visual mode of TinyMCE, event.content is not wrapped in a `<p>` but TinyMCE wraps it anyway.
	- When I switch back from the visual mode to the HTML mode, SaveContent is dispatched and it does not call wp.editor.removep so the `<p>` tags are not automatically removed at that stage.

Now, for the purposes of this test, we explicitly set wpautop to true in the test plugin here: https://github.com/WordPress/gutenberg/blob/3da717b8d0ac7d7821fc6d0475695ccf3ae2829f/packages/e2e-tests/plugins/wp-editor-metabox.php#L36 

This means that we do not want to see the the `<p>` tags in the editor. I added some documentation to that test to explain how to proceed in case it flips back again at some point @youknowriad .

Comment (youknowriad) [reply packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js:56]: Thanks for the detailed explanation :)

**Comment by youknowriad** [comment]
@noisysocks going to backport this one because of issues in e2e tests in wp/5.8 branch

File: packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js (hunk 3 above)
Comment (gziolo) [reply packages/e2e-tests/specs/editor/plugins/wp-editor-meta-box.test.js:56]: Good one @adamziel, this test should be finally reliable 👍🏻
